### ISSUE TITLE
iff AST Node

### ISSUE THREAD

**Issue by bannsec** [JonathanSalwan/Triton#571]
I'm getting an error about iff AST nodes:

```
In [15]: print(rip_ast)
((_ extract 63 0) ((_ zero_extend 0) (ite (= (bvor (bvxor ((_ extract 0 0) ((_ extract 31 31) (bvsub ((_ extract 31 0) ((_ zero_extend 32) SymVar_0)) ((_ sign_extend 24) (_ bv4 8))))) ((_ extract 0 0) ((_ extract 31 31) (bvand (bvxor ((_ extract 31 0) ((_ zero_extend 32) SymVar_0)) ((_ sign_extend 24) (_ bv4 8))) (bvxor ((_ extract 31 0) ((_ zero_extend 32) SymVar_0)) ((_ extract 31 0) (bvsub ((_ extract 31 0) ((_ zero_extend 32) SymVar_0)) ((_ sign_extend 24) (_ bv4 8))))))))) ((_ extract 0 0) (ite (= ((_ extract 31 0) (bvsub ((_ extract 31 0) ((_ zero_extend 32) SymVar_0)) ((_ sign_extend 24) (_ bv4 8)))) (_ bv0 32)) (_ bv1 1) (_ bv0 1)))) (_ bv1 1)) (_ bv20 64) (_ bv5 64))))

In [16]: state.ctx.simplify(rip_ast,True)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-16-743ad7e3a694> in <module>()
----> 1 state.ctx.simplify(rip_ast,True)

TypeError: Z3ToTritonAst::visit(): 'iff' AST node not supported yet
```

I don't see any `iff` operations in there.

**Comment by JonathanSalwan** [comment]
Indeed, in your `rip_ast` expression, there is no `iff` node but using the `simplify` function, z3 simplifies it using such node. So, when Triton converts the z3's simplified expression into the Triton's representation, it raises an exception saying that it doesn't support such nodes.

(duplicate #409)

**Comment by Owlz** [comment]
Gotcha. Is there plans to support this? This actually came up immediately on simplifying rip after a conditional jump, so likely it is something that z3 would use a lot of.

**Comment by JonathanSalwan** [comment]
yep, the problem is that `iff` cannot be supported as a Python representation =/.

**Comment by JonathanSalwan** [comment]
`ite` can be represented as `(0 if x else 1)` but about the `iff` we are fucked :(

**Comment by JonathanSalwan** [comment]
Maybe a potential way should be to ask to z3 to do not use such node during its simplification process (but dunno if it's possible).

**Comment by Owlz** [comment]
So I thought the z3 wrapper just exposed it as an object. Why do we need to be able to express `iff` as a python statement instead of treating it as an opaque object that z3 understands?

**Comment by JonathanSalwan** [comment]
> Why do we need to be able to express iff as a python statement instead of treating it as an opaque object that z3 understands?

Do you have an example?

**Comment by Owlz** [comment]
Just trying to understand the problem. Not sure why Triton has to be able to represent `iff` so long as z3 understands what it means.

**Comment by JonathanSalwan** [comment]
it's ok about z3 but how do we display our AST when `setAstRepresentationMode(PYTHON)` is enabled?

**Comment by Owlz** [comment]
my personal opinion on that is to simply make up some format for it. i would think the benefits of having it optimized would outweigh not having a legit way to print the representation out in python. Also, I wouldn't think anyone would need to directly evaluate the python representation itself. I see that more as a description of the ast (and obviously already leaves out size of the bitvector for example).

**Comment by kenoph** [comment]
What about using an instrinsic? It would break `eval()` but only if you don't import the instrinsic into the global scope.

My whole "I want a Z3 expression in python" issue was specifically because of this issue.

**Comment by JonathanSalwan** [comment]
Notes for myself:

The evaluation part regarding the symbolic emulation:

```smt
; $ z3 -smt2 iff.smt2
; sat
; true
; false
; false
; true

(check-sat)
(eval
    (iff
        (= (_ bv1 16) (_ bv1 16))
        (= (_ bv10 16) (_ bv10 16))
    )
)

(eval
    (iff
        (= (_ bv1 16) (_ bv1 16))
        (= (_ bv20 16) (_ bv10 16))
    )
)

(eval
    (iff
        (= (_ bv2 16) (_ bv1 16))
        (= (_ bv10 16) (_ bv10 16))
    )
)

(eval
    (iff
        (= (_ bv2 16) (_ bv1 16))
        (= (_ bv20 16) (_ bv10 16))
    )
)
```

The Python representation:

```ipython
In [1]: (10 == 10 if 1 == 1 else not (10 == 10))
Out[1]: True

In [2]: (20 == 10 if 1 == 1 else not (20 == 10))
Out[2]: False

In [3]: (10 == 10 if 2 == 1 else not (10 == 10))
Out[3]: False

In [4]: (20 == 10 if 2 == 1 else not (20 == 10))
Out[4]: True
```

![truth_table](https://user-images.githubusercontent.com/991046/51195072-a456a100-18ec-11e9-9881-edb448f0dca5.png)

**Comment by JonathanSalwan** [comment]
Merged into `dev-v0.7`.
